File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        n = len(s)

        if (n%2 == 1) :
            return False
        
        stack = []

        for c in s:
            if(c == '(') :
                stack.append(')')
            elif (c == '[') :
                stack.append(']')
            elif (c == '{') :
                stack.append('}')

            else :
                if not stack or stack.pop() != c :
                    return False

        return not stack

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        int n = s.length() ;

        if (n%2 == 1) return false;

        stack<char> st;

        for( char c : s) {
            if(c == '(')        st.push(')');
            else if (c=='[')    st.push(']');
            else if (c == '{')  st.push('}');

            else {
                if (st.empty() || st.top() != c) return false;
                st.pop();
            }
        }
        return st.empty();
    }
};